core-bundle/src/Resources/contao/library/Contao/Template.php

```diff
@@ -395,6 +398,50 @@ public function trans($strId, array $arrParams=array(), $strDomain='contao_defau
 		return System::getContainer()->get('translator')->trans($strId, $arrParams, $strDomain);
 	}
 
+	/**
+	 * Adds JSON-LD data to the current page.
+	 */
+	public function jsonLd(array $jsonLd, string $schema = JsonLdManager::SCHEMA_ORG): void
+	{
+		$container = System::getContainer();
+		$responseContext = $container->get(ResponseContextAccessor::class)->getResponseContext();
+
+		if (!$responseContext || !$responseContext->has(JsonLdManager::class))
+		{
+			return;
+		}
+
+		/** @var JsonLdManager $jsonLdManager */
+		$jsonLdManager = $responseContext->get(JsonLdManager::class);
+
+		$type = $jsonLdManager->createTypeFromArray($jsonLd);
+		$id = $jsonLd['identifier'] ?? Graph::IDENTIFIER_DEFAULT;
+
+		$jsonLdManager
+			->getGraphForSchema($schema)
+			->set($type, $id);
+	}
+
+	/**
+	 * Finalizes the JSON-LD data collected during the current response context
+	 * and returns the scripts ready to be output within HTML context.
+	 */
+	public function finalizeJsonLd(): string
```

Comment:
I would prefer removing the `Template::finalizeJsonLd()` method and set the script string explicitly in `PageRegular` to something like `$this->Template->jsonLd` (we do the same/similar with the `HtmlHeadBag` IMO).

Reply:
the idea was that this could be relevant for other controllers as well, e.g. an ESI fragment should render its own JSON-LD

Reply:
And shouldn’t the fragment controller then itself just call `$responseContext->get(JsonLdManager::class)->collectFinalScriptFromGraphs()` and give it to its template?

Reply:
No the template defines where it wants to output the JSON-LD.

Reply:
Agree, it should define *where* to output the script, but it schould not generate it itself IMO.

Reply:
Well it's just a function to make things lazy. The same as we use in the `news` for `text` (https://github.com/contao/contao/blob/4.x/news-bundle/src/Resources/contao/modules/ModuleNews.php#L124) for example. Otherwise it would always be generated even if not used. Maybe it's just the name or the fact that it's a function call that puzzles you? What about `<?= $this->jsonLdScripts ?>`?

Reply:
What I don’t like, is that it is a method on the Template class an accesses the “global” container to get and generate data.

> The same as we use in the `news` for `text`

There the news module sets (or injects) the `text` property to the template. If we do this in `PageRegular` with something like `$this->template->jsonLdScripts = function () {…}` I would be totally happy ☺️

Reply:
But then everybody needs to copy that code?

Reply:
They have to create the response context themselves, and therefore should know if it has a JsonLdManager and it would only be a single line of code then:

```
$responseContext->get(JsonLdManager::class)->collectFinalScriptFromGraphs()
```

Reply:
That would look like c40216ffa28991da2087fc8cbcaca45c8fd6d335 then. I don't mind, is that okay with you too @aschempp ?

Reply:
We could pass `$responseContext` as a parameter to `createFooterScripts()` so that we don’t have to ask the container. But I’m fine with how it is now too. 🎉

Reply:
We can't because they were extensible. But I've fixed that with 6664322671c4c34587b015118413b2a18b2a653c now.

Reply:
approved ✅